src/oneD/StFlow.cpp

```diff
@@ -344,118 +359,138 @@ void StFlow::updateProperties(size_t jg, double* x, size_t jmin, size_t jmax)
             }
-
-            // The default boundary condition for species is zero flux. However,
-            // the boundary object may modify this.
-            double sum = 0.0;
-            for (size_t k = 0; k < m_nsp; k++) {
-                sum += Y(x,k,0);
-                rsd[index(c_offset_Y + k, 0)] =
-                    -(m_flux(k,0) + rho_u(x,0)* Y(x,k,0));
+        } else { // interior points
+            if (m_usesLambda) {
+                // Note that this propagates the mass flow rate information to the left
+                // (j+1 -> j) from the value specified at the right boundary. The
+                // lambda information propagates in the opposite direction.
+                rsd[index(c_offset_U,j)] =
+                    -(rho_u(x,j+1) - rho_u(x,j))/m_dz[j]
+                    -(density(j+1)*V(x,j+1) + density(j)*V(x,j));
+            } else if (m_isFree) {
+                // terms involving V are zero as V=0 by definition
+                if (grid(j) > m_zfixed) {
+                    rsd[index(c_offset_U,j)] =
+                        - (rho_u(x,j) - rho_u(x,j-1))/m_dz[j-1];
+                } else if (grid(j) == m_zfixed) {
+                    if (m_do_energy[j]) {
+                        rsd[index(c_offset_U,j)] = (T(x,j) - m_tfixed);
+                    } else {
+                        rsd[index(c_offset_U,j)] = (rho_u(x,j)
+                                                    - m_rho[0]*0.3); // why 0.3?
+                    }
+                } else if (grid(j) < m_zfixed) {
+                    rsd[index(c_offset_U,j)] =
+                        - (rho_u(x,j+1) - rho_u(x,j))/m_dz[j];
+                }
+            } else {
+                // unstrained with fixed mass flow rate
+                rsd[index(c_offset_U, j)] = rho_u(x, j) - rho_u(x, j - 1);
             }
-            rsd[index(c_offset_Y + leftExcessSpecies(), 0)] = 1.0 - sum;
+            diag[index(c_offset_U, j)] = 0; // Algebraic constraint
+        }
+    }
+}
 
-            // set residual of poisson's equ to zero
-            rsd[index(c_offset_E, 0)] = x[index(c_offset_E, j)];
-        } else if (j == m_points - 1) {
-            evalRightBoundary(x, rsd, diag, rdt);
-        } else { // interior points
-            evalContinuity(j, x, rsd, diag, rdt);
-            // set residual of poisson's equ to zero
-            rsd[index(c_offset_E, j)] = x[index(c_offset_E, j)];
 
-            //------------------------------------------------
-            //    Radial momentum equation
-            //
-            //    \rho dV/dt + \rho u dV/dz + \rho V^2
-            //       = d(\mu dV/dz)/dz - lambda
-            //-------------------------------------------------
+//! Evaluate the momentum equation residual.
+void StFlow::evalMomentum(double* x, double* rsd, int* diag,
+                          double rdt, size_t jmin, size_t jmax)
+{
+    for (size_t j = jmin; j <= jmax; j++) {
+        if (j == 0) { // left boundary
```

Comment:
This is the original code for the interior points of the momentum residual equation.

I believe this is correct. We are talking about the momentum residual equation? I think you'll find the remaining logic is located in the Lambda residual equation.

Reply:
The (radial) momentum equation isn’t solved at all if `m_usesLambda` is false. There may be some implicit / undefined behavior.

Reply:
Is this a flaw in the original logic?

I based my adjustments on the following from the original code. Below are all the locations where the momentum equation is touched in the original code.

```

 if (j == 0) {
   rsd[index(c_offset_V,0)] = V(x,0);
} else if (j == m_points - 1) {
   rsd[index(c_offset_V,j)] = V(x,j);
   diag[index(c_offset_V,j)] = 0;
} else {
    if (m_usesLambda) {
        rsd[index(c_offset_V,j)] =
            (shear(x, j) - lambda(x, j) - rho_u(x, j) * dVdz(x, j)
            - m_rho[j] * V(x, j) * V(x, j)) / m_rho[j]
            - rdt * (V(x, j) - V_prev(j));
        diag[index(c_offset_V, j)] = 1;
    } else {
        rsd[index(c_offset_V, j)] = V(x, j);
        diag[index(c_offset_V, j)] = 0;
    }
}

```

And this is the current momentum equation:

```

void StFlow::evalMomentum(double* x, double* rsd, int* diag,
                          double rdt, size_t jmin, size_t jmax)
{
    for (size_t j = jmin; j <= jmax; j++) {
        if (j == 0) { // left boundary
            rsd[index(c_offset_V,0)] = V(x,0);
        } else if (j == m_points - 1) { // right boundary
            rsd[index(c_offset_V,j)] = V(x,j);
            diag[index(c_offset_V,j)] = 0;
        } else { // interior points
            if (m_usesLambda) {
                rsd[index(c_offset_V,j)] =
                    (shear(x, j) - lambda(x, j) - rho_u(x, j) * dVdz(x, j)
                    - m_rho[j] * V(x, j) * V(x, j)) / m_rho[j]
                    - rdt * (V(x, j) - V_prev(j));
                diag[index(c_offset_V, j)] = 1;
            } else {
                rsd[index(c_offset_V, j)] = V(x, j);
                diag[index(c_offset_V, j)] = 0;
            }
        }
    }
}

```

Reply:
Likely an oversight in the original logic. In any case, this should be more efficient / easier to follow. Note that I reordered the code so there's no need to use logic inside of the loops.
```C++
void StFlow::evalMomentum(double* x, double* rsd, int* diag,
                          double rdt, size_t jmin, size_t jmax)
{
    if (m_usesLambda) {
        for (size_t j = jmin; j <= jmax; j++) {
            rsd[index(c_offset_V, j)] = V(x, j);
            diag[index(c_offset_V, j)] = 0;
        }
        return;
    }

    size_t j0 = std::max(jmin, 1);
    size_t j1 = std::min(jmax, m_points - 1);  
    for (size_t j = j0; j < j1; j++) { // interior points
        rsd[index(c_offset_V,j)] = 
                    (shear(x, j) - lambda(x, j) - rho_u(x, j) * dVdz(x, j)
                    - m_rho[j] * V(x, j) * V(x, j)) / m_rho[j]
                    - rdt * (V(x, j) - V_prev(j));
        diag[index(c_offset_V, j)] = 1;
    }
    if (jmin == 0) { // left boundary
        rsd[index(c_offset_V, 0)] = V(x, 0);
        diag[index(c_offset_V, 0)] = 0;
    }
    if (jmax == m_points - 1) { // right boundary
        rsd[index(c_offset_V, jmax)] = V(x, jmax);
        diag[index(c_offset_V, jmax)] = 0;
    }
}
```

Reply:
I do like flattening the nesting. One thing that I've been thinking about are all these little if-else clauses that dictate what to do with the residual if X, Y, or Z is true. And how having them peppered throughout tends towards creating these complex methods that have say 3 or 4 branches that may or may not execute and anyone looking at the method has to mentally dig through them to find the default behavior.

I was thinking towards the two-point control PR and how this feature introduces more if statements in the residual equations.

Reply:
I haven't looked at the 2-point method: I was mainly trying to let loops run without having internal logic.

I think that `evalContinuity` will give the best preview of what to expect when this gets a little more complex.

Reply:
I see now what you're talking about with the logic being out of loops. That definitely would help with the issue I was thinking of.

Reply:
Staring at this further, you could actually move some of the logic into `StFlow::eval`, where you call `StFlow::evalMomentum` if `m_usesLambda` is true, and otherwise a new function
```C++
void StFlow::evalUnused(double* x, double* rsd, int* diag,
                        double rdt, size_t jmin, size_t jmax, size_t offset)
{
    for (size_t j = jmin; j <= jmax; j++) {
        rsd[index(offset, j)] = x[index(offset, j)];
        diag[index(offset, j)] = 0;
    }
}
```
with the extra parameter `offset` being `c_offset_V` (in this case).

This is also the same code as the one used by `evalElectricField` (or `evalLambda` if `m_usesLambda` is `false`), with different offsets.

Reply:
My bad: looks like you need
```C++
    size_t j0 = std::max<size_t>(jmin, 1);
    size_t j1 = std::min(jmax, m_points - 1);
```
(note the `<size_t>` template argument. Make sure that this compiles on your machine ...

Reply:
I'm really struggling with the continuity equation and finding a clean way to split that one up so that the logic is clear to someone viewing it.

Reply:
There are three main branches that correspond to `free-flow`, `unstrained-flow` and `axisymmetric-flow` (this is mostly in the user-facing API, see `DomainFactory` (the internal C++ code just uses Boolean flags). You could split it along these lines.

Reply:
Here's a suggestion (note that the upper index is actually `size_t j1 = std::min(jmax, m_points - 2);`, i.e. I had a bug in the other code I posted):
```C++
void StFlow::evalContinuity(double* x, double* rsd, int* diag,
                            double rdt, size_t jmin, size_t jmax)
{
    if (jmin == 0) { // left boundary
        rsd[index(c_offset_U, 0)] =
            - (rho_u(x,1) - rho_u(x,0)) / m_dz[0]
            - density(1)*V(x,1) + density(0)*V(x,0);
        diag[index(c_offset_U, 0)] = 0; // Algebraic constraint
    }
    if (jmax == m_points - 1) { // right boundary
        if (m_usesLambda) {
            rsd[index(c_offset_U, jmax)] = rho_u(x, jmax);
        } else {
            rsd[index(c_offset_U, jmax)] = rho_u(x, jmax) - rho_u(x, jmax-1);
        }
        diag[index(c_offset_U, jmax)] = 0; // Algebraic constraint
    }

    size_t j0 = std::max<size_t>(jmin, 1);
    size_t j1 = std::min(jmax, m_points - 2);
    if (m_usesLambda) { // "axisymmetric_flow"
        // Note that this propagates the mass flow rate information to the left
        // (j+1 -> j) from the value specified at the right boundary. The
        // lambda information propagates in the opposite direction.
        for (size_t j = j0; j <= j1; j++) {
            rsd[index(c_offset_U,j)] =
                -(rho_u(x,j+1) - rho_u(x,j)) / m_dz[j]
                -(density(j+1)*V(x,j+1) + density(j)*V(x,j));
            diag[index(c_offset_U, j)] = 0; // Algebraic constraint
        }
    } else if (m_isFree) { // "free-flow"
        for (size_t j = j0; j <= j1; j++) {
            // terms involving V are zero as V=0 by definition
            if (grid(j) > m_zfixed) {
                rsd[index(c_offset_U,j)] = -(rho_u(x,j) - rho_u(x,j-1)) / m_dz[j-1];
            } else if (grid(j) == m_zfixed) {
                if (m_do_energy[j]) {
                    rsd[index(c_offset_U, j)] = T(x,j) - m_tfixed;
                } else {
                    rsd[index(c_offset_U, j)] = rho_u(x,j) - m_rho[0]*0.3; // why 0.3?
                }
            } else { // grid(j) < m_zfixed
                rsd[index(c_offset_U, j)] = -(rho_u(x,j+1) - rho_u(x,j)) / m_dz[j];
            }
            diag[index(c_offset_U, j)] = 0; // Algebraic constraint
        }
    } else { // "unstrained-flow" (fixed mass flow rate)
        for (size_t j = j0; j <= j1; j++) {
            rsd[index(c_offset_U, j)] = rho_u(x,j) - rho_u(x,j-1);
            diag[index(c_offset_U, j)] = 0; // Algebraic constraint
        }
    }
}
```

Reply:
Ah, I will try this out. Currently with the incorrect indices, the python test suite freezes my virtual machine haha.

Reply:
This business with setting the diag variable...is there a default 0 value and we only need to set it to 1 where a time-derivative is desired?